### ISSUE TITLE
Import type dont recognize the type and returns any

### ISSUE THREAD

**Issue by crisys11** [flowtype/flow-for-vscode#23]
When i import a type from another file, dont recognize the type:
# Code

./typefile.js

``` js
// @flow
export type ReduxFormField = {
  value: string,
  active: boolean,
  autofill: Function,
  autofilled: ?boolean,
  checked: ?boolean,
  dirty: boolean,
  error: string,
  initialValue: any,
  invalid: boolean,
  name: string,
  onBlur(valueOrEvent: any): Function,
  onChange(valueOrEvent: any): Function,
  onDragStart: Function,
  onDrop: Function,
  onFocus:Function,
  onUpdate: Function,
  pristine: boolean,
  touched: boolean,
  valid: boolean,
  visited: boolean,
};
```

``` js
// @flow
import type {ReduxFormField} from './typefile';
let a: ReduxFormField;
```

autocomplete recognize var "a" with "any" type

Notes
- Nuclide recognize the type fine.
- flow dont throw errors with the import

**Comment by Arilas** [comment]
Can you specify version of VSCode and flow version?

**Comment by crisys11** [comment]
flow 0.30.0
vscode 1.4

**Comment by warmbowski** [comment]
I get the same thing with the same version of flow 0.30.0 and vscode 1.4.0 (extension version is 0.0.3). I also see that starting vscode with this extension doesn't start up the flow server and there are no errors. But even I already have flow running via command line or nuclide, I still have this 'imported types' problem in vscode. So even if flow's running, I don't thing the extension is connecting to the flow server.

Without being able to connect to the flow server, I don't see how vscode would be able to reconcile the imported flow types.

**Comment by satya164** [comment]
Facing the same issue. Happy to PR if someone could point me in the right direction.

**Comment by clooth** [comment]
Same issue here too.

**Comment by ggregoire** [comment]
Same issue with vscode 1.5.3 and flow 0.33.

**Comment by elloyd-spredfast** [comment]
Same issue with VSCode 1.7.2 and flow 0.36. Extension version 0.4.0.

**Comment by orta** [comment]
Hrm, I'm not seeing this when I do this:

![screen shot 2016-12-29 at 3 49 23 pm](https://cloud.githubusercontent.com/assets/49038/21554580/aac830fe-cdde-11e6-9f72-0d75954944dd.png)

Then I get autocomplete as expected:

![screen shot 2016-12-29 at 3 52 39 pm](https://cloud.githubusercontent.com/assets/49038/21554600/dd355f12-cdde-11e6-8a3b-df69994d2402.png)

admittedly I don't get it if I use your exact case of `let a: ReduxFormField` as Flow doesn't seem to imply its a real object at that point.

**Comment by nfour** [comment]
Same issue with the below:

```js
// @flow
import type { StorageEntry } from './StrategyStorage';

export default class Strategy {
  constructor(strategy: StorageEntry) {
    strategy;
  }
}
```

From the file:

```js

export type StorageEntry = {
  Path: string,
  Content: {},
  Attributes: {},
};

```

**Comment by natew** [comment]
I get this too on 1.8.1

**Comment by jcreamer898** [comment]
Similar issue as above...

![](http://d.pr/i/tWHZ+)

If I comment out the import and put the type inline it works...

![](http://d.pr/i/avek+)

**Comment by sattaman** [comment]
+1

**Comment by natew** [comment]
@crisys11 reason for closing?

**Comment by crisys11** [comment]
Works well in vs code 1.9.1 and plugin 0.5.0

**Comment by aweary** [comment]
I'm still seeing this issue with VSCode 1.9.1 and version 0.5.0 of the plugin.

**Comment by cakecatz** [comment]
I think there has been a misunderstanding.
![000](https://cloud.githubusercontent.com/assets/6136383/23112709/cd92d9c0-f775-11e6-8a90-7a798e3b989f.png)
`let a: any` is `Automatic Typing Acquisition`.

**Comment by im-sven** [comment]
Anyone has found a workaround?

**Comment by Arilas** [comment]
@im-sven are you sure, that flow extension is correctly installed and configured? Do you see type hinting like on screenshot in section [Flow], or you see only automatic intellisense info with 'any'. 

Check that file have correct header. Maybe you need to use `/** @flow*/` header

Check language that used for file(on the right bottom line). It's must be just 'Javascript'. 

If you still have this issue, I think that this is something related to VSCode itself, and his resolver autocomplete resolver.

**Comment by im-sven** [comment]
@Arilas thanks for your comment. After starting a completely fresh project, I couldn't reproduce the behaviour. It seemed that it was caused by an incorrect `flowconfig`. 

I suspect it was caused by using the `modules.resolve` in Webpack, to import files like: 
```
import foo from 'folder1/subfolder'
```
and not handling this correctly in Flow. In my case (for illustrational purposes), it was fixed by updating the .flowconfig` as:
```
module.name_mapper='^\(folder1\|folder2\|folder3\|folder4\)\/\(.*\)$' -> '<PROJECT_ROOT>/editor/'
module.system.node.resolve_dirname=node_modules
```

Hope it can point other people in the right direction.

**Comment by Leeds-eBooks** [comment]
I'm not using webpack and I'm still seeing this issue.

**Comment by MoeSattler** [comment]
Same problem here. First time user, just freshly installed. it is not working.

**Comment by Arilas** [comment]
@Leeds-eBooks @MoeSattler Can you specify what exactly don't work for you? Is it hover and/or autocomplete? Because it's related to different APIs of VSCode.

**Comment by Leeds-eBooks** [comment]
@Arilas It's both actually. When I work out how to do a screenshot of a hover effect I will paste it here, but basically if I `import type {T} from 'path'` hovering over `T` shows `any` and then, when referencing a variable that has type `T` and calling a property, the autocomplete doesn't have any of the type's properties.

**Comment by MoeSattler** [comment]
Both for me too

**Comment by Arilas** [comment]
Do you have flow annotation for file with types? Do you have global nodejs installation or use nvm or similar tools. Do you have globally installed flow-bin or just dev dependency? Did you configure `flow.pathToFlow`?

**Comment by Leeds-eBooks** [comment]
>Do you have flow annotation for file with types?

Yes, `import type {T} from 'path'` imports from a file with flowtype annotations.

>Do you have global nodejs installation or use nvm or similar tools.

I use nvm. But in any case, flow-for-vscode is working fine in its other features.

>Do you have globally installed flow-bin or just dev dependency?

Globally installed.

>Did you configure flow.pathToFlow?

No.

**Comment by Kureev** [comment]
Have exactly the same issue. Flow version 0.38, plugin version 0.5.0. `import Foo from './types'` works, but `import type {Foo} from './types'` doesn't. 

types.js:
```js
export type Foo = {bar: number};
```

**Comment by Kureev** [comment]
Ok, I can't pretend I know the root cause, but(!):
- If you're using `import type { Foo } from './types'` then type is actually imported and can be used, but at the moment of import it'll be displayed as `any` (which I think is a bug).  
- If you're using imported types for prop validations, they won't be displayed:
```js
class Bar extends React.Component {
  props: {
    foo: Foo
  }

  render() {
    const { bar } = this.props.foo;
  }
}
```
`foo` will be displayed as `any` here, but `bar` will have a proper type.

Is it something we can help you fixing, @Arilas?

**Comment by AbsoluteZero273** [comment]
I am having this issue too.

**Comment by srgcpr** [comment]
In my case I just forgot to add `/* @flow */` in file with type declarations. Little mistake but maybe for someone who's googling why imported types are not working will be helpful.

**Comment by fignuts** [comment]
I'm also having this issue. Flow 0.45.0 vscode 1.12.1.

**Comment by ButuzGOL** [comment]
and also reference jumping broken

**Comment by rossPatton** [comment]
i am having a very similar problem which might be related to this

i have a large project, and I use flow-typed along with a lot of custom types. i define it in my .flowconfig like so:

```
[libs]
flow-typed/
```

my work folder is defined in the same file like this:

```
[include]
src/
```

Now, somewhere deep in `src/` I have a react component i'm working on. That component has some props defined like this:

```
type Props = {
  selectedAge?: tSelectedAge;
};
```

`tSelectedAge` is a union type defined somewhere in `/flow-typed`

The flow cli tool has 0 problem with this setup. This seems to be the standard way to share library definitions among the flow community. However, when I hover over `selectedAge` it gets a type of `any`. So `flow-for-vscode` is not picking up 

Here's where it's weird. If I open on one of my type declaration files in `flow-typed`, any of the types in that file now magically propagate to all my other files in my workspace

If I have, say, `flow-typed/_globals.js` open, where I define my `tSelectedAge` type, then `selectedAge` in my react component is now typed properly as a string union instead of `any`

My `.flowconfig` is in the root of my project, but it seems like maybe this extension is not picking it up? As is, it makes the extension kind of unusable because it completely throws off the coverage stats and misses a lot of errors

**Comment by ryangomba** [comment]
Hoping to help here. This didn't work for me:

```js
export type Foo = {
    bar: string,
};
// ...
module.exports = baz;
```

But this did:

```js
export type Foo = {
    bar: string,
};
// ...
export default baz;
```

**Comment by foobar8675** [comment]
I had the same issue and got it working using default props, as outlined https://wietse.loves.engineering/using-flowtype-with-decorators-in-react-af4fe69e66d6

**Comment by HappyZombies** [comment]
Spent maybe 30 mins on this and well...it was really simple. Should've read this issue comments more closely... 

If you have a types.js file, you must add the `// @flow` annotation on your types.js file as well. Else you will get this odd behavior.

**Comment by jmcrthrs** [comment]
@rossPatton I experience the same. Opening any declaration files cause the types to appear in autocomplete. As soon as I close a declaration file, they disappear from autocomplete.

Flow Language Support: 0.7.0
Flow: 0.48.0
VS Code: 1.13.1

**Comment by liushigit** [comment]
My setup:
 flow-bin: 0.53.1
VSCode: 1.15.1

`.flowconfig`

```
[ignore]
.*/node_modules/config-chain/test/broken.json
.*/node_modules/react-event-listener/src/index.js
[include]
[libs]
[options]
```
I have `"flow.useNPMPackagedFlow": true` in my VS Code settings.

The type def below is in a `types.js`
```
...
export type AppState = {
  cities: string[],
  currentCity: string,
  infosByCity: InfosByCity
};
```

It is imported as below...

![s1](https://user-images.githubusercontent.com/5442413/29742018-1a1e8100-8aaa-11e7-9bde-3c5e871fdd4e.png)
![s2](https://user-images.githubusercontent.com/5442413/29742017-1a1d8a98-8aaa-11e7-8118-087666993bdc.png)

**Comment by DarrylD** [comment]
@liushigit any luck? having the same issue.

**Comment by michaellee8** [comment]
Any luck, i have the same issue in atom as well

**Comment by malekjaroslav** [comment]
Do we have some new info on this? It makes using this plugin really annoying :(

**Comment by orta** [comment]
Nope, if it's being seen in Atom, it's likely at the Flow level rather than this plugin. 

If you'd like to get involved I think #150 would be a good place to start, as that lets the Flow team handle more of the work in this extension.

**Comment by greeze** [comment]
Same issue here.

However, I'm noticing that if I change:
`import type { Foo, Bar } from './types'`
to
`import { type Foo, type Bar } from './types'`
it works as expected!

**Comment by t2** [comment]
@greeze What version of `flow`, `flow-bin`, and VS Code are you on? I am not seeing the same results.

**Comment by crisys11** [comment]
actually works for me.

**Comment by greeze** [comment]
@t2 Latest VSCode, `flow-bin@0.61.0` and, interestingly, `flow version` reports `0.54.1`. I'm not sure how to update my `flow` version, though.

**Comment by t2** [comment]
@greeze Do you have `flow.useNPMPackagedFlow` set to true or false?

**Comment by greeze** [comment]
@t2 It defaulted to `false` but changing it to `true` doesn't seem to resolve the issue.

**Comment by crisys11** [comment]
please check both files types and implementation have flow anotation

**Comment by greeze** [comment]
Both files have `// @flow` as their first line.

**Comment by t2** [comment]
@crisys11 @greeze That led me down the path to fixing it all. You must have `// @flow` in every file that you want to import as well. I am golden. Thanks for the help!

**Comment by patrickgordon** [comment]
I have `// flow` on all files but still I have imported types showing as "Any".

@t2 could you show a code sample / screenshot of this working?

**Comment by t2** [comment]
@patrickgordon You have `// flow` or // @flow`? Here is a screenshot of it working https://cl.ly/0C1r0j1J2w18. When I hover over the `import` I still see any https://cl.ly/321C2r1T0K1T but the types are being checked correctly in the actual usage. 

flow-bin: 0.61.0

**Comment by patrickgordon** [comment]
Sorry, I have `// @flow`.
I will try bumping my flow version see if that helps

**Comment by patrickgordon** [comment]
Hmm no luck. Really stumped.

**Comment by t2** [comment]
@patrickgordon Try these things.

Extensions https://cl.ly/0N0w2I3L0h3z
Flow Language Support Setting: https://cl.ly/092w3Q2Y0l0W
Homebrew installed Flow 0.61.0 as well

**Comment by patrickgordon** [comment]
OK so it looks like if I hover over the alias type that uses the imported type it doesn't work but when looking at the actual JS that references the type it does. That's a win.

**Comment by t2** [comment]
@patrickgordon Yeah, that's what I am seeing. Not ideal but functional.

**Comment by Izhaki** [comment]
@greeze answer worked for me:

    import { type Foo, type Bar } from './types'

People still experience this issue?

**Comment by sanghin** [comment]
> People still experience this issue?

Yes i still have the issue

OS: **macOS High Sierra 10.13.2 (17C88)**
VSCode: **Version 1.19.2 (1.19.2)**
Extensions: (**should i list all my extensions ?** )
![image](https://user-images.githubusercontent.com/3429798/35277894-87b1365e-0048-11e8-8290-8813fd75837b.png)

VSCode **Workspace settings**:

```js
{
  "javascript.validate.enable": false,

  "flow.pathToFlow": "${workspaceRoot}/node_modules/.bin/flow",
  "flow.useNPMPackagedFlow": true,
  "flow.fileExtensions": [".js"],
  "flow.runOnAllFiles": true,

  "prettier.eslintIntegration": true,
  "search.exclude": {
    "**/node_modules": true,
    "**/bower_components": true,
    "**/tmp/**": true
  }
}
```

VSCode **User settings** :
```js
{
  "sublimeTextKeymap.promptV3Features": true,
  "editor.multiCursorModifier": "ctrlCmd",
  "editor.snippetSuggestions": "top",
  "window.title": "${activeEditorLong}${separator}${rootName}",
  "window.zoomLevel": 1,
  "editor.minimap.enabled": false,
  "editor.tabCompletion": true,
  "editor.tabSize": 2,
  "workbench.editor.enablePreview": false,
  "typescript.autoImportSuggestions.enabled": false,
  "editor.parameterHints": false,
  "openInGitHub.requireSelectionForLines": false,
  "openInGitHub.providerType": "github",
  "editor.formatOnSave": true,
  "editor.formatOnPaste": true,
  "editor.formatOnType": true,
}
```
AddonType.js
```js
// @flow

export type Addon = (
  id: string
  type: 'free' | 'pro'
};
```

WorkspaceType.js
```js
// @flow

import { type Addon } from "./AddonType";
// import { type Addon } from "./AddonType";

export type Workspace = {
  id: string,
  addons: Addon[]
};
```

`With import type { Addon } from "./AddonType";`
![image](https://user-images.githubusercontent.com/3429798/35279576-baf08c18-004d-11e8-835b-6f3c8fa803f8.png)

`With import { type Addon } from "./AddonType"`
![image](https://user-images.githubusercontent.com/3429798/35279450-6816e8f2-004d-11e8-85bf-7f9ef40c8467.png)

**Comment by Izhaki** [comment]
I'm not using vscode-flow-ide, perhaps try without? @sanghin

**Comment by sanghin** [comment]
It's good now

**TL;DR don't export your type with `export * from './File';`**

I was exporting the type like this (in index.js file):

```javascript
export * from "./AddonType";
export * from "./WorkspaceType";
[...]
```

Where i should export it like this:

```javascript
// @flow

import type {
   FlowIsGreatAddons,
  TestAddons,
  Addons,
} from "./AddonType";
import type { Workspace } from "./WorkspaceType";

export type {
   FlowIsGreatAddons,
  TestAddons,
  Addons,
  Workspace
};
```

Import =>

```javascript
import type { User, Subscription, Workspace } from '@core/AppTypes';
// or without alias
import type { User, Subscription, Workspace } from '../../AppTypes';
```

Hovering on the type return `any`:
![image](https://user-images.githubusercontent.com/3429798/35286827-452e60a4-0061-11e8-93ba-0cd09995586b.png)

But hovering on the JS files display the types and Flow now display errors if i tried to access a non existing props, etc.

As describe here:

> OK so it looks like if I hover over the alias type that uses the imported type it doesn't work but when looking at the actual JS that references the type it does. That's a win.

Thank tou for your time (and sorry for wasting it 🙃  )

**Comment by romanlex** [comment]
Same issues(
Doesn't work...all types any(

**Comment by TwFlem** [comment]
@romanlex, @sanghin comment will address the issue. Also, it will work using export *. Make sure you are adding // @flow to ALL FILES INVOLVED. This involves where they are defined, being exported, being used, and anywhere in-between.

**Comment by romanlex** [comment]
@TwFlem with export * all work, thx!

**Comment by vschoener** [comment]
Same issue, doesn't work :(

**Comment by sanghin** [comment]
Make sure that:

- you are adding // @flow to ALL FILES INVOLVED. 

Tried to import the definition directly from your file, don't use a index.js as a 'proxy'.

packages/files.js:
```js
type stuff = Array<string>;
export {
  stuff 
};
```

```js
import type { stuff } from 'packages/files.js'
```

*AND NOT*

```js
import type { stuff } from 'packages';
```

**Comment by Khachatour** [comment]
Hey there all.
I somehow managed to get this working.

`import {type MyType} from 'whereIExportType.js'`

**Comment by Leeds-eBooks** [comment]
@Khachatour Of course, that is fine. I think this issue is fixed now anyway?

**Comment by wilsonshakespeare** [comment]
> Same issue here.
> 
> However, I'm noticing that if I change:
> `import type { Foo, Bar } from './types'`
> to
> `import { type Foo, type Bar } from './types'`
> it works as expected!

Ya it does but then it would have problem with `flow-remove-types/register` because it doesn't recognize the latter

**Comment by stepankuzmin** [comment]
Same here.

vs code 1.28.2
flow-for-vscode 0.7.2
flow: 0.85.0

`.flowconfig`

```
[ignore]
.*/node_modules/findup/.*
.*/node_modules/@mapbox/jsonlint-lines-primitives/.*

[include]

[libs]
./flow-typed
./node_modules/mapbox-gl/flow-typed

[options]
esproposal.class_static_fields=enable
esproposal.class_instance_fields=enable

[lints]
```

`flow-typed/types.js`

```js
// @flow

import type Map from 'mapbox-gl/src/ui/map';

declare type MapboxMap = Map;
```

This resolves correctly:

<img width="466" alt="screenshot 2018-10-31 13 08 47" src="https://user-images.githubusercontent.com/533564/47781178-1f58e400-dd0e-11e8-8a8a-bc5f3ab39fd5.png">

But everywhere in the project it resolves to any in VS Code:

<img width="334" alt="screenshot 2018-10-31 13 05 34" src="https://user-images.githubusercontent.com/533564/47780983-ace80400-dd0d-11e8-9974-49dd2d97a3c7.png">

Although if I run flow on source code it works fine.

**Comment by ButuzGOL** [comment]
I still have issue with such way 
```
import type {
  CombatType,
  WarriorType,
} from '../lib/types';
```
but dont have with 
```
import {
  type CombatType,
  type WarriorType,
} from '../lib/types';
```
But what is right way to do ?

This doc doesn't give right example
https://flow.org/en/docs/types/modules/#importing-and-exporting-types-

In google i see different syntax in articles

**Comment by seanyesmunt** [comment]
What fixed it for me, and how I think I should have been doing it in the first place(?) was to move all my type files into the `/flow-typed` folder. Then changed `export type` to `declare type`.

Then I don't need to manually import all of my types, they just work.
https://flow.org/en/docs/libdefs/

**Comment by falnos24865** [comment]
This really should not have been closed. Its still an issue and renders this extension almost useless if using a code base that does imports the way the flow docs specify - `import type Foo, {MyObject, MyInterface} from './exports';`

**Comment by EAT-CODE-KITE-REPEAT** [comment]
+1

**Comment by ortonomy** [comment]
For anyone stumbling upon this, suggest checking out this thread and this comment: https://github.com/facebook/flow/issues/338#issuecomment-349634841

**Comment by demkovych** [comment]
Still happens with me

**Comment by pizza-r0b** [comment]
![image](https://user-images.githubusercontent.com/5282495/65174396-ad60ba80-da05-11e9-8770-5011e1dda961.png)

This still seems to happen. In this screenshot I am hovering over form.exit - you can see that the proper type is displayed in the tool tip on hover, but under that is `any`. Why is any there? It's typed, so it should not be any

**Comment by BrunoQuaresma** [comment]
I have the same issue.

**Comment by Mayank1791989** [comment]
@BrunoQuaresma Is it still happening. Share some sample code?

**Comment by BrunoQuaresma** [comment]
@Mayank1791989 yeap.

**Comment by Mayank1791989** [comment]
I think any is coming from typescript plugin not from this plugin.

**Comment by Qhkm92** [comment]
> @greeze answer worked for me:
> 
> ```
> import { type Foo, type Bar } from './types'
> ```
> 
> People still experience this issue?

This works for me

**Comment by GCDeveloper** [comment]
@seanyesmunt answer worked for me. I am using flow-typed. This was solved by moving my types exports into the flow-typed folder (in my case `flow-types/types/GeometryTypes.js`), changing `export` to `declare`, and finally removing `import`.

**Comment by lportanel** [comment]
What worked for me was:
`// @flow
import { type Foo, type Bar } from './types'`

And very important: don't forget to add `// @flow` to the file where you're exporting your types.
